connect/runtime/src/test/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerderTest.java

```diff
@@ -726,21 +683,21 @@ public void testAccessors() throws Exception {
         doNothing().when(connectorConfigCb).onCompletion(any(NotFoundException.class), isNull());
 
         // Create connector
-        connector = mock(BogusSourceConnector.class);
+        Connector connector = mock(BogusSourceConnector.class);
```

Comment:
This is also correct, and is more consistent with the other test cases.
```suggestion
        Connector connector = mock(SourceConnector.class);
```
After you do this, you can change expectConfigValidation to take a SourceSink instead of a Connector, and call the mock(SinkConnector.class) or mock(SourceConnector.class) internally. The calling test doesn't ever need to see the connector variable, because that object is only ever needed for config validation in the main code.

Reply:
Agreed on the connector being created internally, but wouldn't that mean that we only need the config argument? I think I missed the usage of the new SourceSink variable in your explanation

Reply:
You would have expectConfigValidation accept a sourceSink, and then basically do this:
```
Connector connectorMock = sourceSink == SourceSink.SOURCE ? mock(SourceConnector.class) : mock(SinkConnector.class);
```

Reply:
Sorry for being late here, had a few hospital visits.
that makes sense 👍 . For tests like `testPutConnectorConfig` where the function is called twice, the test fails as the mock is different for each function call. I'm yet to debug why having the same mock is the right way but thought I'd ask you first

Reply:
I just verified that it's actually the reason for the tests failing. I can adjust expectConfigValidation to return the mocked connector and reuse it? or would that defeat the purpose of the mock not being visible to the test?

Reply:
> For tests like testPutConnectorConfig where the function is called twice, the test fails as the mock is different for each function call. I'm yet to debug why having the same mock is the right way but thought I'd ask you first

It would help if I actually told you what this mock is doing. The strange two-calls, boolean argument, multiple configs is related to just this one computeIfAbsent call: https://github.com/apache/kafka/blob/159d25a7df25975694e2e0eb18a8feb125f7c39e/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java#L755

Each time the `AbstractHerder` (superclass of `StandaloneHerder`) runs a validation and processes a config it calls `computeIfAbsent`, and only calls newConnector once. This single Connector instance then validates every configuration for that worker.

So the first call to `expectConfigValidation` (with argument true) mock out the creation, and the following calls (with argument false) just add more config invocations to the existing mocked object. If you create a new mocked object, those are never used because the original mocked object is re-used.

Rather than returning the mocked object, I think it would be better that a single expectConfigValidation call mocked out all of the configs in-advance, like `testCreateConnectorAlreadyExists` does with two `config` arguments. The `Map` argument in the `false` calls can be moved to the `true` calls.

Reply:
Thanks for the walkthrough! I now understand the full picture. I think I addressed all your comments now, let me know if I missed anything